src/common/prop.c

```diff
@@ -471,28 +490,78 @@ daos_prop_copy(daos_prop_t *prop_req, daos_prop_t *prop_reply)
 out:
 	if (rc) {
 		if (label_alloc) {
-			entry_req = daos_prop_entry_get(prop_req,
-							DAOS_PROP_PO_LABEL);
-			D_FREE(entry_req->dpe_str);
+			free_str_prop_entry(prop_req, DAOS_PROP_PO_LABEL);
+			free_str_prop_entry(prop_req, DAOS_PROP_CO_LABEL);
 		}
 		if (acl_alloc) {
-			entry_req = daos_prop_entry_get(prop_req,
-							DAOS_PROP_PO_ACL);
-			D_FREE(entry_req->dpe_val_ptr);
+			free_ptr_prop_entry(prop_req, DAOS_PROP_PO_ACL);
+			free_ptr_prop_entry(prop_req, DAOS_PROP_CO_ACL);
 		}
 		if (owner_alloc) {
-			entry_req = daos_prop_entry_get(prop_req,
-							DAOS_PROP_PO_OWNER);
-			D_FREE(entry_req->dpe_str);
+			free_str_prop_entry(prop_req, DAOS_PROP_PO_OWNER);
+			free_str_prop_entry(prop_req, DAOS_PROP_CO_OWNER);
 		}
 		if (group_alloc) {
-			entry_req = daos_prop_entry_get(prop_req,
-						DAOS_PROP_PO_OWNER_GROUP);
-			D_FREE(entry_req->dpe_str);
+			free_str_prop_entry(prop_req, DAOS_PROP_PO_OWNER_GROUP);
+			free_str_prop_entry(prop_req, DAOS_PROP_CO_OWNER_GROUP);
 		}
 		if (entries_alloc) {
 			D_FREE(prop_req->dpp_entries);
 		}
 	}
 	return rc;
 }
+
+
+int
+daos_prop_entry_dup_ptr(struct daos_prop_entry *entry_dst,
+			struct daos_prop_entry *entry_src, size_t len)
+{
+	D_ASSERT(entry_src != NULL);
+	D_ASSERT(entry_dst != NULL);
+
+	D_ALLOC(entry_dst->dpe_val_ptr, len);
+	if (entry_dst->dpe_val_ptr == NULL)
+		return -DER_NOMEM;
+
+	memcpy(entry_dst->dpe_val_ptr, entry_src->dpe_val_ptr, len);
+	return 0;
+}
+
+int
+daos_prop_entry_cmp_acl(struct daos_prop_entry *entry1,
+			struct daos_prop_entry *entry2)
+{
+	struct daos_acl *acl1;
+	struct daos_acl *acl2;
+
+	/* Never call this with entries not known to be ACL types */
+	D_ASSERT(entry1->dpe_type == DAOS_PROP_PO_ACL ||
+		 entry1->dpe_type == DAOS_PROP_CO_ACL);
+	D_ASSERT(entry2->dpe_type == DAOS_PROP_PO_ACL ||
+		 entry2->dpe_type == DAOS_PROP_CO_ACL);
+
+	if (entry1->dpe_val_ptr == NULL && entry2->dpe_val_ptr == NULL)
+		return 0;
+
+	if (entry1->dpe_val_ptr == NULL || entry2->dpe_val_ptr == NULL) {
+		D_ERROR("ACL mismatch, NULL ptr\n");
+		return -DER_MISMATCH;
+	}
+
+	acl1 = entry1->dpe_val_ptr;
+	acl2 = entry2->dpe_val_ptr;
+
+	if (acl1->dal_len != acl2->dal_len) {
+		D_ERROR("ACL len mistmatch, %u != %u\n",
+			acl1->dal_len, acl2->dal_len);
+		return -DER_MISMATCH;
+	}
+
+	if (memcmp(acl1, acl2, daos_acl_get_size(acl1)) != 0) {
```

Comment:
Why use daos_acl_get_size here but not in the conditional check for length above?

Reply:
daos_acl_get_size includes the header size and the ACE len, whereas dal_len just includes the ACE len. In practice the comparison result would be the same either way since the ACE len is the only piece that varies in size.

I figured for the actual memcmp, we'd want to compare the whole structure, not just the entries. Not a lot of extra info in the header, but it includes the version, for example.

Reply:
I went ahead and changed this to make it more consistent.